projects/geometry/101_color_cube_metal/101_color_cube_metal.cpp

```diff
@@ -0,0 +1,253 @@
+    if (vsShader.Function.get() == nullptr) {
+        assert(false && "VS Shader MTL::Library::newFunction() failed");
+        return EXIT_FAILURE;
+    }
+
+    fsShader.Function = NS::TransferPtr(library->newFunction(NS::String::string("fragmentMain", NS::UTF8StringEncoding)));
+    if (fsShader.Function.get() == nullptr) {
+        assert(false && "FS Shader MTL::Library::newFunction() failed");
+        return EXIT_FAILURE;
+    }
+
+    // *************************************************************************
+    // Graphics pipeline state object
+    // *************************************************************************
+    MetalPipelineRenderState renderPipelineState;
+    MetalDepthStencilState   depthStencilState;
+    CHECK_CALL(CreateDrawVertexColorPipeline(
+        renderer.get(),
+        &vsShader,
+        &fsShader,
+        GREX_DEFAULT_RTV_FORMAT,
+        GREX_DEFAULT_DSV_FORMAT,
+        &renderPipelineState,
+        &depthStencilState));
+
+    // *************************************************************************
+    // Geometry data
+    // *************************************************************************
+    MetalBuffer pIndexBuffer;
+    MetalBuffer pPositionBuffer;
+    MetalBuffer pVertexColorBuffer;
+    CreateGeometryBuffers(renderer.get(), &pIndexBuffer, &pPositionBuffer, &pVertexColorBuffer);
+
+    // *************************************************************************
+    // Window
+    // *************************************************************************
+    auto window = Window::Create(gWindowWidth, gWindowHeight, "101_color_cube_metal");
+    if (!window) {
+        assert(false && "Window::Create failed");
+        return EXIT_FAILURE;
+    }
+
+    // *************************************************************************
+    // Render Pass Description
+    // *************************************************************************
+    MTL::RenderPassDescriptor* pRenderPassDescriptor = MTL::RenderPassDescriptor::renderPassDescriptor();
+
+    // *************************************************************************
+    // Swapchain
+    // *************************************************************************
+    if (!InitSwapchain(renderer.get(), window->GetNativeWindow(), window->GetWidth(), window->GetHeight(), 2, MTL::PixelFormatDepth32Float)) {
+        assert(false && "InitSwapchain failed");
+        return EXIT_FAILURE;
+    }
+
+    // *************************************************************************
+    // Main loop
+    // *************************************************************************
+    MTL::ClearColor clearColor(0.23f, 0.23f, 0.31f, 0);
+    uint32_t        frameIndex = 0;
+
+    while (window->PollEvents()) {
+        CA::MetalDrawable* pDrawable = renderer->Swapchain->nextDrawable();
+
+        // nextDrawable() will return nil if there are no free swapchain buffers to render to
```

Comment:
Does `nextDrawable()` wait at all?

Reply:
It does, by default setAllowableTimeout() is true, which waits for 1 second. As we talked about I'm going to assert if it doesn't give back a drawable as after a second it probably means some error has occurred.